calyx-opt/src/passes/top_down_compile_control.rs

```diff
@@ -212,6 +212,27 @@ fn compute_unique_ids(con: &mut ir::Control, cur_state: u64) -> u64 {
     }
 }
 
+/// Given the state of the FSM, returns the index for the register in `fsms``
+/// that should be queried.
+/// A query for each state must read from one of the `num_registers` registers.
+/// For `r` registers and `n` states, we split into "buckets" as follows:
+/// `{0, ... , n/r - 1} -> reg. @ index 0`,
+/// `{n/r, ... , 2n/r - 1} -> reg. @ index 1`,
+/// ...,
+/// `{(r-1)n/r, ... , n - 1} -> reg. @ index n - 1`.
+/// Note that dividing each state by the value `n/r`normalizes the state w.r.t.
+/// the FSM register from which it should read. We can then take the floor of this value
+/// (or, equivalently, use unsigned integer division) to get this register index.
+fn register_to_query(state: u64, num_states: u64, num_registers: u64) -> usize {
+    // num_states+1 is needed to prevent error (the done condition needs
+    // to check past the number of states, i.e., will check fsm == 3 when
+    // num_states == 3).
+    let reg_to_query: usize = (state * num_registers / (num_states + 1))
```

Comment:
So I had a whole back and forth with myself about this `+1`, and I eventually got rid of it for the following reason. Please lmk if my reasoning is off:

When you use `register_to_query`, you pass in `fsm_rep.last_state`, which is itself computed by `sch.last_state()`. This function looks through all the transitions, which are tuples of `u64`s, and returns the maximum `u64` value along the _second_ element of the tuple. As you pointed out in the comment, this includes the state above the index of the true last state, meaning `sch.last_state()` returns the number of states, and not the index of the last state. I think the way it currently is, with the `+ 1`, only the final state will query from the second register because that's the way the integer division works out if the denominator `num_states` is too high. Lmk if that checks out.

Reply:
The reason why I did `+1` was only pragmatic: 
If you have a single register, then you always want `reg_to_query` to return 0. The only way to do this for the last state (i.e., when `state==num_state` is if you do `num_states +1`). 

The only thing i don't quite understand is this: 
```
I think the way it currently is, with the + 1, only the final state will query from the second register because that's the way the integer division works out if the denominator num_states is too high. 
```
Won't it query from the second register any time `state > (num_states + 1)/2` which will be basically half of the states?